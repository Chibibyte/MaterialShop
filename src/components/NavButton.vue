<template>
  <v-btn v-if="routerPush" @click="routerNav" :disabled="disabled">
    <v-icon :left="showText">{{ icon }}</v-icon>
    <span v-if="showText">{{ name }}</span>
  </v-btn>
  <router-link v-else :to="to" v-slot="{ href, route, navigate }">
    <v-btn
      :disabled="disabled"
      :href="href"
      @click="(e) => (navigate ? navigate(e) : console.log('no navigate:', e))"
    >
      <v-icon :left="showText">{{ icon }}</v-icon>
      <span v-if="showText">{{ name }}</span>
    </v-btn>
  </router-link>
</template>

<script>
/**
 * Business Logic
 *    -button for navigation within this single page application
 *    -needs router
 *    -for navigation without reload
 *    -can run other function before routing
 *
 * Inputs:
 *    -to: routing target
 *    -icon: material icon name
 *    -name: button name
 *    -disabled: toggle button
 *    -routerPush: if set, routing with router.push; else router-link
 *    -runBefore: function that runs before routing
 * Outputs:
 *    -routerPush
 *    -disabled
 *    -icon
 *    -name
 */
export default {
  name: "NavButton",
  computed: {
    showText() {
      return !this.$vuetify.breakpoint.xs;
    },
  },
  props: {
    to: {
      type: String,
      default: "/",
    },
    icon: {
      type: String,
      default: "",
    },
    name: {
      type: String,
      default: "",
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    routerPush: {
      type: Boolean,
      default: false,
    },
    runBefore: Function,
  },
  methods: {
    routerNav() {
      if (this.runBefore) this.runBefore();
      this.$router.push(this.to);
    },
  },
};
</script>