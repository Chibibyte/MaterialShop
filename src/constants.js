export default {
    BUTTONICONS: {
        AUTHENTICATION: 'mdi-login-variant',
        ADDRESS: 'mdi-package-variant',
        BANKACCOUNT: 'mdi-bank',
        SHIPPING: 'mdi-truck',
        PAYMENT: 'mdi-cash',
        SUMMARY: 'mdi-text-box-check-outline',
        CHECKOUT_COMPLETE: 'mdi-check',
        ADD: 'mdi-plus',
        REMOVE: 'mdi-minus',
        CANCEL: 'mdi-window-close',
        CONFIRM: 'mdi-check',
        SHOPPING: 'mdi-shopping',
        SUCCESS: 'mdi-check',
        FAILURE: 'mdi-window-close',
        WARNING: 'mdi-alert-circle-outline'
    }
}