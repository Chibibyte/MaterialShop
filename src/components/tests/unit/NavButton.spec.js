import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils';
import NavButton from '@/components/NavButton.vue';
import constants from '@/constants.js'
import VueRouter from 'vue-router'

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueRouter);
localVue.prototype.$constants = constants;
let router = new VueRouter();

let stubs = [
    'v-btn',
    'v-icon',
]

let wrapper;
let mocks;
beforeEach(() => {
    mocks = {
        // $router: { push: jest.fn() }
    };
    wrapper = shallowMount(NavButton, { stubs, mocks, localVue, router });
})

describe('Component', () => {
    test('is instantiated', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    })

    test('default snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    })
})

describe('props:runBefore', () => {
    test('runBefore:routerPush == false', async () => {
        await wrapper.setProps({ runBefore: jest.fn() });
        wrapper.vm.routerNav();
        expect(wrapper.vm.runBefore.mock.calls).toHaveLength(1);
        expect(wrapper.element).toMatchSnapshot();
    })

    test('runBefore:routerPush == true', async () => {
        await wrapper.setProps({ runBefore: jest.fn(), routerPush: true });
        wrapper.vm.routerNav();
        expect(wrapper.vm.runBefore.mock.calls).toHaveLength(1);
        expect(wrapper.element).toMatchSnapshot();
    })
})
