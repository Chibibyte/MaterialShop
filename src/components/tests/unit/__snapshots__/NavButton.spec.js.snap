// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component default snapshot 1`] = `
<router-link-stub
  ariacurrentvalue="page"
  event="click"
  tag="a"
  to="/"
>
  <v-btn-stub>
    <v-icon-stub
      left=""
    >
      
    </v-icon-stub>
     
    <span>
      
    </span>
  </v-btn-stub>
</router-link-stub>
`;

exports[`props:runBefore runBefore:routerPush == false 1`] = `
<router-link-stub
  ariacurrentvalue="page"
  event="click"
  tag="a"
  to="/"
>
  <v-btn-stub>
    <v-icon-stub
      left=""
    >
      
    </v-icon-stub>
     
    <span>
      
    </span>
  </v-btn-stub>
</router-link-stub>
`;

exports[`props:runBefore runBefore:routerPush == true 1`] = `
<v-btn-stub>
  <v-icon-stub
    left=""
  >
    
  </v-icon-stub>
   
  <span>
    
  </span>
</v-btn-stub>
`;
